programs/storage_api/src/storage_contract.rs

```diff
@@ -45,3 +54,303 @@ pub enum StorageContract {
+            reward_validations,
+            lockout_validations,
+        } = &mut storage_contract
+        {
+            let original_segments = *state_entry_height / ENTRIES_PER_SEGMENT;
+            let segments = entry_height / ENTRIES_PER_SEGMENT;
+            debug!(
+                "advertise new last id segments: {} orig: {}",
+                segments, original_segments
+            );
+            if segments <= original_segments {
+                return Err(InstructionError::InvalidArgument);
+            }
+
+            *state_entry_height = entry_height;
+            *state_hash = hash;
+
+            // move lockout_validations to reward_validations
+            *reward_validations = lockout_validations.clone();
+            lockout_validations.clear();
+            lockout_validations.resize(segments as usize, Vec::new());
+            self.set_state(storage_contract)
+        } else {
+            Err(InstructionError::InvalidArgument)?
+        }
+    }
+
+    fn proof_validation(
+        &mut self,
+        entry_height: u64,
+        proofs: Vec<CheckedProof>,
+        replicator_accounts: &mut [KeyedAccount],
+    ) -> Result<(), InstructionError> {
+        let mut storage_contract = &mut self.state()?;
+        if let StorageContract::Default = storage_contract {
+            *storage_contract = StorageContract::ValidatorStorage {
+                entry_height: 0,
+                hash: Hash::default(),
+                lockout_validations: vec![],
+                reward_validations: vec![],
+            };
+        };
+
+        if let StorageContract::ValidatorStorage {
+            entry_height: current_entry_height,
+            lockout_validations,
+            ..
+        } = &mut storage_contract
+        {
+            if entry_height >= *current_entry_height {
+                return Err(InstructionError::InvalidArgument);
+            }
+
+            let segment_index = get_segment_from_entry(entry_height);
+            let mut previous_proofs = replicator_accounts
+                .iter_mut()
+                .filter_map(|account| {
+                    account.state().ok().map(move |contract| match contract {
+                        StorageContract::ReplicatorStorage { proofs, .. } => {
+                            Some((account, proofs[segment_index].clone()))
+                        }
+                        _ => None,
+                    })
+                })
+                .flatten()
+                .collect::<Vec<_>>();
+
+            if previous_proofs.len() != proofs.len() {
+                // don't have all the accounts to validate the proofs against
+                return Err(InstructionError::InvalidArgument);
+            }
+
+            let mut valid_proofs: Vec<_> = proofs
+                .into_iter()
+                .enumerate()
+                .filter_map(|(i, entry)| {
+                    if previous_proofs[i].1.signature != entry.proof.signature
```

Comment:
Putting this all in a separate function would let you use early returns, would make this code very short, and more testable.

Reply:
I'll look into it! Thanks!

Another thing, I'm not testing tightly over these functions, instead I'm using a bank and a bank client to test them. Mainly because I want the runtime checks to be performed post execute. It'd be nice if there was a simple way to check that accounts weren't modified in a bad way.

Reply:
Agreed. I played with a bunch of variations, but I think using that BankClient is the best we have to offer at the moment. The interface there is fairly stable, whereas ENTRYPOINT is such a cumbersome interface, gotta figure it's bound to change.